### PR TITLE
WIP: basic udp listener support

### PR DESCRIPTION
**Pull request by cmluciano**
*Description*: Basic UDP listener skeleton and some utility functions
*Risk Level*: Medium 
*Testing*: bazel tests //...
*Docs Changes*: TODO
*Release Notes*: TODO

xref: #492

### CODE REVIEW

**Comment by cmluciano** [comment]
/review @alyssawilk 

I'm still having some problems with the ListenerImplTest. My guess is that the dispatcher or address class is defaulting to creating a stream socket, but I can't figure out where. All I am left with is an error about ```where GetParam() = 4-byte object <00-00 00-00>``` .

```
bazel test //test/common/network:listener_impl_test --test_filter=ListenerImplTest.UdpSetListeningSocketOptionsSuccess:ListenerImplTest/*.UdpSetListeningSocketOptionsSuccess:*/ListenerImplTest.UdpSetListeningSocketOptionsSuccess/*:*/ListenerImplTest/*.UdpSetListeningSocketOptionsSuccess
```

File: source/common/network/listen_socket_impl.cc [hunk 1]
```diff
@@ -60,6 +60,35 @@ TcpListenSocket::TcpListenSocket(int fd, const Address::InstanceConstSharedPtr&
...
+    : ListenSocketImpl(address->socket(Address::SocketType::Datagram), address) {
+  RELEASE_ASSERT(fd_ != -1, "");
+
+  // TODO(htuch): This might benefit from moving to SocketOptionImpl.
+  // TODO(cmluciano): Analyze special options to be enabled/disabled for UDP
+  int on = 1;
+  int rc = setsockopt(fd_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
+  RELEASE_ASSERT(rc != -1, "");
+  rc = setsockopt(fd_, SOL_SOCKET, SO_SNDBUF, &on, sizeof(on));
+  RELEASE_ASSERT(rc != -1, "");
```
Comment (alyssawilk) [inline source/common/network/listen_socket_impl.cc]: useful error messages on these?
RELEASE_ASSERT(rc != -1, fmt::format("failed to set UDP send buffer to {} bytes"));

**Comment by mattklein123** [comment]
> But there's no listen/accept semantics because UDP

Here is how I have been thinking about how UDP will work in general.

1) UDP listeners require a "connection" idle timeout to be set.
2) When a UDP packet is received, the remote IP/port is used to lookup the associated "UDP" connection. If there is one, it is used. If not, one is created w/ the normal filter stack semantics. 
3) UDP "connections" can only be destroyed via idle timeout.

Later on we can work on packet forwarding, worker/"connection" affinity, but I think the first step is to just get ^ working where each worker is independent. 

Does that make sense?

**Comment by alyssawilk** [comment]
To be clear, do you think this is for UDP proxying, stateful UDP, or all UDP?

I think this totally makes sense for UDP proxying (modulo we may eventually want to make the hashing configurable because of the NAT port rebind issue, but we can deal with that later =P)

For QUIC we'll want to be able to have the transport/application layer be able to terminate the underlying connection as it has the context to know when it has been torn down.  Also if we think there's doing to be any state associated with the UDP 'connect' we will need to deal with port migration, which the QUIC layer can deal with by verifying CID and consistent use of keys.  If we define the APIs right it's plausible we can do the usual Envoy registered factory thing and the QUIC session manager can just be a subclass of the generic session manager but we'll need application logic in there somewhere.

For stateless UDP a la DNS I have less of a design sense (done little of that over here) other than the fact we'll want to have really minimal session state per packet since there's no actual association between one packet from [5 tuple] and the next.  As long as the base class UDP 'connection' is super light weight it should be fine but UDP proxying will end up carrying enough state I believe we won't want struct reuse between that and DNS.  When/if we start adding state I guess it can be its own UDP registered factory session subclass too.

**Comment by mattklein123** [comment]
> To be clear, do you think this is for UDP proxying, stateful UDP, or all UDP?

What I wrote above is clearly targeted at stateful UDP, but IMO we can start with it for UDP proxying (e.g. DNS) even though it's heavyweight. I think as you say we can simplify and make the basic UDP proxy case more lightweight in the future probably by completely doing away with a filter stack per "connection" and instead just have a fixed/static filter stack per listener. This would still allow programmability but would be very high performance. Thoughts?

> For QUIC we'll want to be able to have the transport/application layer be able to terminate the underlying connection as it has the context to know when it has been torn down.

I will admit I'm hazier on the QUIC side of things, but my general thinking here is that QUIC is really going to be a hybrid between pure UDP (stateful/proxy) and TCP. Meaning, a QUIC listener might use UDP code under the hood but will spit out connections that are more TCP-like in that they support close, event notifications, etc.?

**Comment by cmluciano** [comment]
@alyssawilk @mattklein123 Any more comments that need addressed before I incorporate the feedback in this PR?

**Comment by mattklein123** [comment]
@cmluciano I haven't fully reviewed. Please just ping when the PR is done and ready for review again?

File: source/common/network/listen_socket_impl.cc [hunk 2]
```diff
@@ -60,6 +60,35 @@ TcpListenSocket::TcpListenSocket(int fd, const Address::InstanceConstSharedPtr&
...
+                                 bool bind_to_port)
+    : ListenSocketImpl(address->socket(Address::SocketType::Datagram), address) {
+  RELEASE_ASSERT(fd_ != -1, "");
+
+  // TODO(htuch): This might benefit from moving to SocketOptionImpl.
+  // TODO(cmluciano): Analyze special options to be enabled/disabled for UDP
+  int on = 1;
+  int rc = setsockopt(fd_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
+  RELEASE_ASSERT(rc != -1, fmt::format("failed to set UDP SO_REUSEADDR socket option"));
+  rc = setsockopt(fd_, SOL_SOCKET, SO_SNDBUF, &on, sizeof(on));
```
Comment (mpwarres) [inline source/common/network/listen_socket_impl.cc]: Maybe I'm misreading, but I think the SO_SNDBUF and SO_RCVBUF sockopts here might be configuring the socket to have send and receive buffers of size 1 (byte).  An alternative might be to leave out these setsockopt calls so the socket uses system default buffer sizes, and if someone wants to tune the buffer sizes, they could do by specifying socket_options in the config for the listener?

Comment (conqerAtapple) [inline source/common/network/listen_socket_impl.cc]: nit: Why not add a `setOption` method in the base `Socket` class like:
```
 template <typename V>
    static int setOption(int fd, int level, int name, V* value) {
        return ::setsockopt(fd, level, name, (void*)value, sizeof(V));
    }
```

Then you can use the method as well as becomes a nice utility method for the `Socket` class.

Comment (cmluciano) [reply source/common/network/listen_socket_impl.cc]: Great catch, this was a poor choice on my part. I will leave these parts out for now and leave a note to shift these into a factory for later.

Comment (mpwarres) [reply source/common/network/listen_socket_impl.cc]: Sounds good, thanks!  BTW, let me know if you'd like an extra pair of hands for this, I'd be happy to help in whatever way is constructive. (Also fully understand if that just complicates things and you'd prefer not--just wanted to put the offer out there.)

Comment (cmluciano) [reply source/common/network/listen_socket_impl.cc]: I like it. I think there are some notes from Harvey to do this for the TCP defaults as well.

Comment (cmluciano) [reply source/common/network/listen_socket_impl.cc]: Happy to have help if I get stuck, will ping you on Slack early next week.

**Comment by mattklein123** [comment]
Marking as waiting. 

/wait

File: source/common/network/listen_socket_impl.cc (hunk 2 above)
Comment (cmluciano) [reply source/common/network/listen_socket_impl.cc]: @htuch Per your comment [here](https://github.com/envoyproxy/envoy/blob/master/source/common/network/listen_socket_impl.cc#L45) . Should this leverage something like [SocketOptionImpl::setSocketOption](https://github.com/envoyproxy/envoy/blob/master/source/common/network/socket_option_impl.cc#L28)

Comment (htuch) [reply source/common/network/listen_socket_impl.cc]: Probably.. certainly going via the OS syscall shim layer is desirable for testing, and when using socket options that are not universally supported, it's nice to have the additional machinery provided there.

**Comment by cmluciano** [comment]
Not stale

**Comment by cmluciano** [comment]
@mattklein123 Yes if someone is willing to pick this up or we can can try to split it up into smaller parallel pieces, I'm all for it. I'm still familiarizing myself with different pieces of the codebase and I don't want to block anyone from making forward progress quicker. I'd still really like to contribute to this new feature even if it just means helping with writing more tests.

**Comment by mattklein123** [comment]
@cmluciano sounds good. @conqerAtapple do you want to pick up this PR? @cmluciano can help with reviews, additional tests, etc. as we get going. Thank you everyone!

**Comment by mattklein123** [comment]
@cmluciano should we close in favor of https://github.com/envoyproxy/envoy/pull/5108?

**Comment by mpwarres** [comment]
I've drafted a PR that fills this out and adds another test at #5115.  Past that, I think additional APIs/code will be needed to handle receiving messages. Here's an [initial proposal](https://docs.google.com/a/google.com/document/d/e/2PACX-1vTvrfAp7v_tMvwKSyq5kfJGCkRBS47kojkOzvTywujjOuqVKsuUPH0mffr9aW9aIrfHd-gAAzkGTZS6/pub).

**Comment by cmluciano** [comment]
Thanks @mpwarres , will take a look today

**Pull request closed without merging**